### PR TITLE
ddl: refactor V2 args for RebaseAutoID/Modify-table/Index-visibility/FK related DDL.

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: Ref https://github.com/pingcap/tidb/issues/53930
Ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:
The ddl types that this pr refactor:
- RebaseAutoID
- RebaseAutoRandomBase
- ModifyTableComment
- ModifyTableCharsetAndCollate
- AlterIndexVisibility
- AddForeignKey
- DropForeignKey

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/meta/model/job_args.go [hunk 1]
```diff
@@ -607,6 +607,199 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
...
+		}, nil
+	}
+
+	// for version V2
+	return getOrDecodeArgsV2[*RebaseAutoIDArgs](job)
+}
+
+// ModifyTableCommentArgs is the arguments for ActionModifyTableComment ddl.
+type ModifyTableCommentArgs struct {
+	Comment string
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:648]: add json tags

File: pkg/meta/model/job_args.go [hunk 2]
```diff
@@ -607,6 +607,199 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
...
+	}
+
+	return getOrDecodeArgsV2[*ModifyTableCommentArgs](job)
+}
+
+// ModifyTableCharsetAndCollateArgs is the arguments for ActionModifyTableCharsetAndCollate ddl.
+type ModifyTableCharsetAndCollateArgs struct {
+	ToCharset          string
+	ToCollate          string
+	NeedsOverwriteCols bool
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:680]: doitto

File: pkg/meta/model/job_args.go [hunk 3]
```diff
@@ -607,6 +607,199 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
...
+		return args, nil
+	}
+
+	return getOrDecodeArgsV2[*ModifyTableCharsetAndCollateArgs](job)
+}
+
+// AlterIndexVisibilityArgs is the arguments for ActionAlterIndexVisibility ddl.
+type AlterIndexVisibilityArgs struct {
+	IndexName pmodel.CIStr
+	Invisible bool
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:709]: all below args doesn't have json tags

File: pkg/meta/model/job_args.go [hunk 4]
```diff
@@ -607,6 +607,199 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
...
+}
+
+// AddForeignKeyArgs is the arguments for ActionAddForeignKey ddl.
+type AddForeignKeyArgs struct {
+	FkInfo  *FKInfo
+	FkCheck bool
+}
+
+func (a *AddForeignKeyArgs) fillJob(job *Job) {
+	intest.Assert(job.Version == JobVersion1 || job.Version == JobVersion2, "job version is invalid")
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:748]: put it in job.FillArgs, no need to add in every args

File: pkg/ddl/foreign_key_test.go [hunk 5]
```diff
@@ -58,34 +58,40 @@ func testCreateForeignKey(t *testing.T, d ddl.ExecutorForTest, ctx sessionctx.Co
...
 	job := &model.Job{
+		Version:    model.GetJobVerInUse(),
 		SchemaID:   dbInfo.ID,
 		SchemaName: dbInfo.Name.L,
 		TableID:    tblInfo.ID,
 		TableName:  tblInfo.Name.L,
 		Type:       model.ActionDropForeignKey,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []any{pmodel.NewCIStr(foreignKeyName)},
+		Args:       []any{},
```
Comment (D3Hunter) [inline pkg/ddl/foreign_key_test.go:89]: why left it here

File: pkg/meta/model/job.go [hunk 6]
```diff
@@ -868,6 +868,7 @@ func (job *Job) GetInvolvingSchemaInfo() []InvolvingSchemaInfo {
 // SubJob is a representation of one DDL schema change. A Job may contain zero
 // (when multi-schema change is not applicable) or more SubJobs.
 type SubJob struct {
+	Version     JobVersion      `json:"version"`
```
Comment (D3Hunter) [inline pkg/meta/model/job.go:871]: can we make all job version related to multi schema change to v1 as we have discussed

Comment (joccau) [reply pkg/meta/model/job.go]: I have checked the correctness of V2 job.args. New I have roll back to V1.

File: pkg/ddl/multi_schema_change.go [hunk 7]
```diff
@@ -251,11 +251,22 @@ func fillMultiSchemaInfo(info *model.MultiSchemaInfo, job *JobWrapper) error {
...
 	case model.ActionRebaseAutoID, model.ActionModifyTableComment, model.ActionModifyTableCharsetAndCollate:
 	case model.ActionAddForeignKey:
-		fkInfo := job.Args[0].(*model.FKInfo)
+		var fkInfo *model.FKInfo
+		if job.Version == model.JobVersion1 {
+			fkInfo = job.Args[0].(*model.FKInfo)
+		} else {
+			args := job.Args[0].(*model.AddForeignKeyArgs)
+			fkInfo = args.FkInfo
+		}
```
Comment (D3Hunter) [inline pkg/ddl/multi_schema_change.go:269]: we can use jobW.JobArgs here

File: pkg/meta/model/job_args_test.go [hunk 8]
```diff
@@ -441,6 +441,83 @@ func TestDropColumnArgs(t *testing.T) {
...
+			args, err := GetRebaseAutoIDArgs(j2)
+			require.NoError(t, err)
+			require.Equal(t, inArgs, args)
+		}
+	}
+}
+
+func TestGetModifyTableCommentArgs(t *testing.T) {
+	inArgs := &ModifyTableCommentArgs{
+		Comment: "TiDb is great",
```
Comment (D3Hunter) [inline pkg/meta/model/job_args_test.go:462]:
```suggestion
		Comment: "TiDB is great",
```

File: pkg/ddl/multi_schema_change.go (hunk 7 above)
Comment (joccau) [reply pkg/ddl/multi_schema_change.go]: Done.

File: pkg/ddl/multi_schema_change.go [hunk 9]
```diff
@@ -251,11 +251,16 @@ func fillMultiSchemaInfo(info *model.MultiSchemaInfo, job *JobWrapper) error {
 		col := job.Args[0].(*table.Column)
 		info.ModifyColumns = append(info.ModifyColumns, col.Name)
 	case model.ActionAlterIndexVisibility:
-		idxName := job.Args[0].(pmodel.CIStr)
+		var idxName pmodel.CIStr
+		if job.Version == model.JobVersion1 {
+			idxName = job.Args[0].(pmodel.CIStr)
+		} else {
+			idxName = job.Args[0].(*model.AlterIndexVisibilityArgs).IndexName
+		}
```
Comment (D3Hunter) [inline pkg/ddl/multi_schema_change.go:259]: doitto

Comment (joccau) [reply pkg/ddl/multi_schema_change.go]: done

**Review by D3Hunter — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Pull request merged**